prov/usnic/src/usdf_eq.c

```diff
@@ -327,60 +276,39 @@ usdf_eq_sread_fd(struct fid_eq *feq, uint32_t *event, void *buf, size_t len,
 
 	eq = eq_ftou(feq);
 
-	/* block awaiting event */
+	/* Setup poll context to block until the FD becomes readable. */
 	pfd.fd = eq->eq_fd;
 	pfd.events = POLLIN;
+
 retry:
 	ret = poll(&pfd, 1, timeout);
-	if (ret < 0) {
+	if (ret < 0)
 		return -errno;
-	} else if (ret == 0) {
+	else if (ret == 0)
 		return -FI_EAGAIN;
-	}
-
-	pthread_spin_lock(&eq->eq_lock);
 
-	/* did someone steal our event? */
-	if (usdf_eq_empty(eq)) {
-		pthread_spin_unlock(&eq->eq_lock);
+	ret = _usdf_eq_read(eq, event, buf, len, flags);
+	if (ret == -FI_EAGAIN)
 		goto retry;
-	}
-
-	if (usdf_eq_error(eq)) {
-		ret = -FI_EAVAIL;
-		goto done;
-	}
-
-	/* consume the event in eventfd */
-	ret = read(eq->eq_fd, &val, sizeof(val));
-	if (ret != sizeof(val)) {
-		ret = -FI_EIO;
-		goto done;
-	}
-
-	ret = usdf_eq_read_event(eq, event, buf, len, flags);
 
-done:
-	pthread_spin_unlock(&eq->eq_lock);
 	return ret;
 }
 
-static ssize_t
-usdf_eq_write_fd(struct fid_eq *feq, uint32_t event, const void *buf,
-		size_t len, uint64_t flags)
+ssize_t usdf_eq_write_internal(struct usdf_eq *eq, uint32_t event,
+		const void *buf, size_t len, uint64_t flags)
 {
-	struct usdf_eq *eq;
-	uint64_t val;
+	uint64_t val = 1;
 	int ret;
 	int n;
 
-	USDF_DBG_SYS(EQ, "\n");
-
-	eq = eq_ftou(feq);
+	USDF_DBG_SYS(EQ, "event=%#" PRIx32 " flags=%#" PRIx64 "\n", event,
+			flags);
 
 	pthread_spin_lock(&eq->eq_lock);
 
-	/* EQ full? */
+	/* Return -FI_EAGAIN if the EQ is full.
+	 * TODO: Disable the EQ.
+	 */
 	if (atomic_get(&eq->eq_num_events) == eq->eq_ev_ring_size) {
```

Comment:
So here is where the queue disabling should probably happen. I was thinking I'd include that in a separate PR, but I could throw a third commit on this. I was thinking I'd add a `disabled` member in the queue. Whenever we encounter a situation where we can't enqueue anymore, then we set the state of the queue to `disabled`. All writes after that fail with `-FI_EOPBADSTATE` and the reads will continue returning events as normal until the end where it will return `-FI_EOPBADSTATE`. How does that sound @jsquyres?